proposals/0000-asset-primitive.md

```diff
@@ -0,0 +1,233 @@
+
+*TODO: More drawbacks? Please contribute!*
+
+# Alternatives
+
+All viable, proposed alternatives can be built on top of this primitive as high-level features. This section outlines the alternatives, and then illustrates how this RFC could power them. The goal is to show that this RFC is not at odds with future, high-level features that make this primitive easier to work with.
+
+## The Assets Folder
+
+An alternative approach that has been considered is to have a special `assets/` folder that is a sibling to `public/` and `src/`. Astro would include every asset in the final build, and then give you a way to look up those assets dynamically at runtime. 
+
+This RFC is not at odds with this alternative. If this alternative were proposed, it would make sense to take advantage of the Asset primitive outlined here and build that high-level feature on top of this low-level primitive, vs. re-implementing Vite features that already exist in Vite (`import.meta.glob`, resolving and loading hooks, etc).
+
+```js
+/* Example: A user-land assets/ folder lookup implementation
+// src/assets/index.js
+export const assets = transformPathToName(import.meta.globEager('./**'));
+export function asset(name) { return assets[name] };
+
+// anywhere else in your project:
+import {assets} from '../assets/index.js';
+<Icon use={assets['book']} />
+
+// or, an alternative API
+import {asset} from '../assets/index.js';
+<Icon use={asset('book')} />
+```
+
+```js
+/* Example: A core assets/ folder lookup implementation (for illustration only) */
+Astro.asset = function(name: string) {
+  const allAssets = import.meta.globEager('/assets/**');
+  const path = getPathFromName(name);
+  return allAssets[path];
+}
+```
+
+We could even make the magic assets folder an official concept in Astro, and allow for a completely, zero-JS asset syntax for the user:
+
+```astro
+---
+/* Example: An icon component that knows about the assets folder */
+/*          <Icon name="book" /> */
+const iconSvg = (await import(`~assets/${Astro.props.name}.svg`)).default;
+--- 
+<Fragment set:html={iconSvg} />
+```
+
+
+## A sugar `local:use` directive
+
+We already have an accepted RFC for the `local:src` and `local:href` directives, which is sugar for the following:
+
+```
+// Sugar:
+<img local:src="../assets/book.png" />
+// Equivilent to this:
+<img src={await import('../assets/book.png?url')} />
+```
+
+That RFC and this RFC both build on the same core idea: We need to pass assets through Vite for them to be included in the final build. You could imagine that we could extend this `local:src` syntax to handle other assets, building on this same primitive:
```

Comment:
I don't believe `local:` is scoped to only src and href at the moment. The accepted RFC doesn't mention that: https://github.com/withastro/rfcs/blob/main/proposals/0011-relative-url-scheme.md

My understanding is that the right-hand side is not special-cased in any way, so I would expect a `local:use` to work once that proposal is completed.

Reply:
good catch, clarified